src/library/scala/util/matching/Regex.scala

```diff
@@ -276,7 +276,7 @@ class Regex private[matching](val pattern: Pattern, groupNames: String*) extends
     case null => None
     case _    =>
       val m = pattern matcher s
-      if (runMatcher(m)) Some((1 to m.groupCount).toList map m.group)
+      if (runMatcher(m)) Some(List.tabulate(m.groupCount)(i => m.group(i + 1)))
```

Comment:
We could extract `Some(List.tabulate(m.groupCount)(i => m.group(i + 1)))` into a private helper method to avoid some duplication.

You could also special case `m.groupCount == 0` and return `Some(Nil)` to avoid the cost of allocating the closure and `ListBuilder`.

Reply:
cant we just eliminate the closure generation and the ListBuilder completely by using a List directly, and building in reverse, as m.group is not sequential. That way no special case as well, and it should be minorly faster and allocate a few bytes less memory

ie 
```
private def groupsList(m:Matcher) = {
  var res = List.empty[String]
  var index = m.groupCount
  while (index >0 ) {
     res ::= m.group(index)
     index  -= 1
  }
  res
}
```
separately (as a separate PR) tabulate should change to remove the builder on the special case is in https://github.com/rorygraves/scalac_perf/issues/26

Reply:
Hi @mkeskells @retronym 

I tested your new suggestions:

```
==== jason

[info] # Run complete. Total time: 00:04:08
[info] 
[info] Benchmark                                                                    Score     Error   Units
[info] RegexUnapplyBenchmark.t8022CharSequence                                    586.808 ?  89.729   ns/op
[info] RegexUnapplyBenchmark.t8022CharSequence:?gc.alloc.rate                     602.439 ?  74.070  MB/sec
[info] RegexUnapplyBenchmark.t8022CharSequence:?gc.alloc.rate.norm                544.001 ?   0.002    B/op
[info] RegexUnapplyBenchmark.t8022Match                                           625.507 ?  29.606   ns/op
[info] RegexUnapplyBenchmark.t8022Match:?gc.alloc.rate                           1503.461 ?  77.124  MB/sec
[info] RegexUnapplyBenchmark.t8022Match:?gc.alloc.rate.norm                      1476.001 ?  10.691    B/op
[info] RegexUnapplyBenchmark.t8787nullMatch                                         3.024 ?   0.126   ns/op
[info] RegexUnapplyBenchmark.t8787nullMatch:?gc.alloc.rate                          0.001 ?   0.002  MB/sec
[info] RegexUnapplyBenchmark.t8787nullMatch:?gc.alloc.rate.norm                    ? 10??              B/op
[info] RegexUnapplyBenchmark.t8787nullMatcher                                    2263.867 ? 219.190   ns/op
[info] RegexUnapplyBenchmark.t8787nullMatcher:?gc.alloc.rate                     1193.695 ?  92.480  MB/sec
[info] RegexUnapplyBenchmark.t8787nullMatcher:?gc.alloc.rate.norm                4216.004 ?   0.007    B/op

[info] # Run complete. Total time: 00:03:10
[info] 
[info] Benchmark                                                                       (groupCount)     Score     Error   Units
[info] RegexUnapplyGroupsBenchmark.groupingBenchmark                                              1   129.875 ?  22.154   ns/op
[info] RegexUnapplyGroupsBenchmark.groupingBenchmark:?gc.alloc.rate                               1  1123.459 ? 132.734  MB/sec
[info] RegexUnapplyGroupsBenchmark.groupingBenchmark:?gc.alloc.rate.norm                          1   224.000 ?   0.001    B/op
[info] RegexUnapplyGroupsBenchmark.groupingBenchmark                                             10   811.480 ?  41.711   ns/op
[info] RegexUnapplyGroupsBenchmark.groupingBenchmark:?gc.alloc.rate                              10   753.748 ?  35.794  MB/sec
[info] RegexUnapplyGroupsBenchmark.groupingBenchmark:?gc.alloc.rate.norm                         10   960.001 ?   0.003    B/op
[info] RegexUnapplyGroupsBenchmark.groupingBenchmark                                            100  9697.988 ? 152.338   ns/op
[info] RegexUnapplyGroupsBenchmark.groupingBenchmark:?gc.alloc.rate                             100   605.034 ?   9.324  MB/sec
[info] RegexUnapplyGroupsBenchmark.groupingBenchmark:?gc.alloc.rate.norm                        100  9240.016 ?   0.032    B/op


====mike

[info] # Run complete. Total time: 00:04:11
[info] 
[info] Benchmark                                                                    Score     Error   Units
[info] RegexUnapplyBenchmark.t8022CharSequence                                    980.078 ? 402.243   ns/op
[info] RegexUnapplyBenchmark.t8022CharSequence:?gc.alloc.rate                     405.827 ? 110.270  MB/sec
[info] RegexUnapplyBenchmark.t8022CharSequence:?gc.alloc.rate.norm                544.001 ?   0.003    B/op
[info] RegexUnapplyBenchmark.t8022Match                                           700.702 ?  64.061   ns/op
[info] RegexUnapplyBenchmark.t8022Match:?gc.alloc.rate                           1350.310 ? 101.211  MB/sec
[info] RegexUnapplyBenchmark.t8022Match:?gc.alloc.rate.norm                      1477.309 ?  11.524    B/op
[info] RegexUnapplyBenchmark.t8787nullMatch                                         3.135 ?   0.119   ns/op
[info] RegexUnapplyBenchmark.t8787nullMatch:?gc.alloc.rate                          0.001 ?   0.002  MB/sec
[info] RegexUnapplyBenchmark.t8787nullMatch:?gc.alloc.rate.norm                    ? 10??              B/op
[info] RegexUnapplyBenchmark.t8787nullMatcher                                    2243.569 ?  97.682   ns/op
[info] RegexUnapplyBenchmark.t8787nullMatcher:?gc.alloc.rate                     1195.593 ?  47.183  MB/sec
[info] RegexUnapplyBenchmark.t8787nullMatcher:?gc.alloc.rate.norm                4216.004 ?   0.008    B/op



[info] # Run complete. Total time: 00:03:11
[info] 
[info] Benchmark                                                                       (groupCount)     Score     Error   Units
[info] RegexUnapplyGroupsBenchmark.groupingBenchmark                                              1   125.933 ?   5.712   ns/op
[info] RegexUnapplyGroupsBenchmark.groupingBenchmark:?gc.alloc.rate                               1  1131.994 ?  49.202  MB/sec
[info] RegexUnapplyGroupsBenchmark.groupingBenchmark:?gc.alloc.rate.norm                          1   224.000 ?   0.001    B/op
[info] RegexUnapplyGroupsBenchmark.groupingBenchmark                                             10   752.294 ?   6.785   ns/op
[info] RegexUnapplyGroupsBenchmark.groupingBenchmark:?gc.alloc.rate                              10   810.403 ?   7.000  MB/sec
[info] RegexUnapplyGroupsBenchmark.groupingBenchmark:?gc.alloc.rate.norm                         10   960.001 ?   0.002    B/op
[info] RegexUnapplyGroupsBenchmark.groupingBenchmark                                            100  8929.270 ? 235.613   ns/op
[info] RegexUnapplyGroupsBenchmark.groupingBenchmark:?gc.alloc.rate                             100   657.388 ?  17.414  MB/sec
[info] RegexUnapplyGroupsBenchmark.groupingBenchmark:?gc.alloc.rate.norm                        100  9240.015 ?   0.030    B/op
```

Looks like Mike's proposal is slight more performant. I think we can sacrifice some readability to adopt this proposal as this is library code. Are you happy with the latest change, Jason?